src/Bicep.Cli/Services/CompilationService.cs

```diff
@@ -47,20 +47,38 @@ public class CompilationService
             this.decompiler = decompiler;
         }
 
-        public async Task RestoreAsync(string inputPath)
+        public async Task RestoreAsync(string inputPath, bool forceModulesRestore)
         {
             var inputUri = PathHelper.FilePathToFileUrl(inputPath);
             var configuration = this.configurationManager.GetConfiguration(inputUri);
             var sourceFileGrouping = SourceFileGroupingBuilder.Build(this.fileResolver, this.moduleDispatcher, this.workspace, inputUri, configuration);
             var originalModulesToRestore = sourceFileGrouping.ModulesToRestore;
 
+            // Ignore modules to restore logic, include all modules to be restored
+            var allModules = sourceFileGrouping.SourceFilesByModuleDeclaration
```

Comment:
>allModules [](http://example.com/codeflow?start=16&length=10)

`allModules` gets calculated even if the user is not forcing, but it's only used with force.

Reply:
You're right, I wanted this to be more readable for the line after but it gets calculated all the time.

Reply:
My other worry with doing the calculations like this is that the logic could get out of sync with the logic called by `Build()`.

Could we just pass a flag into `Build()` so it ignores restore status and always acts like the module is not restored?

Reply:
Ok, so the idea would be to pass the `forceModulesRestore` bool value all the way down to `bicep/src/Bicep.Core/Workspaces/SourceFileGroupingBuilder.cs / private SourceFileGrouping Build()`... and down to `PopulateRecursive()` to centralize the logic for `ModuleRestoreStatus`? Down there we would force a value of `ModuleRestoreStatus.Unknown` if we see that `forceModulesRestore` is `true`?

I thought about this. It would be cleaner if `SourceFileGroupingBuilder.Build() / PopulateRecursive()` return all modules if `forceModulesRestore`. I was hesitant to go that route but you're probably right that the modules status / replacement logic should go somewhere centralized when we ask for a forced restore.

Reply:
Would we stop in `SourceFileGroupingBuilder` or go even deeper into `ModuleDispatcher.GetModuleRestoreStatus()` to completely override the status returned if asked for a forced restore? My bet would be down to `GetModuleRestoreStatus`.

Reply:
TBH, I originally thought about this while coding the feature, centralizing the modules to restore override logic, and told myself that it could be a future refactor. But you're right, why wait.

I did it this morning and It should bring a better separation of concern.

Reply:
@majastrz I changed how it works a bit and It should bring a better separation of concern. Let me know what you think of having a new ModuleDispatcher that is used only when a force restore is requested.

I will add test if you like this, just wanted to validate before writing tests more test for the new ModuleDispatcher

Reply:
Could we add a flag that makes the following true for all module refs?
<img width="668" alt="image" src="https://user-images.githubusercontent.com/22460039/163909367-46ee1f20-a967-4a13-b89a-73044da54626.png">

Then, we wouldn't need any special handling for forced restore and could avoid the extra logging. The decision would be made by the same component as today and nothing would need to change in CLI. And we wouldn't need a separate dispatcher.

Reply:
It is a possibility to propagate a forceRestore value down to `PopulateRecursive()` and short circuit the `restoreStatus` to `Unknown` when asked to force restore. It would be the quickest way to patch this.

IMO SourceFileGroupingBuilder and PopulateRecursive() are starting to get more responsibilities that they should. Maybe a decorator approach at the ModuleDispatcher level is too wide for our need. We could have the logic of ForceRestoreModuleDispatcher in a new interface to overrides only the RestoreModules & GetModuleRestoreStatus logic?

A bit like this:

```c#

// In CLI / CompilationService.cs

        public async Task RestoreAsync(string inputPath, bool forceModulesRestore)
        {
            var inputUri = PathHelper.FilePathToFileUrl(inputPath);
            var configuration = this.configurationManager.GetConfiguration(inputUri);

            if(forceModulesRestore) {
                this.moduleDispatcher.SetRestoreHandler(new ForceModuleRestoreHandler())
            }

            var sourceFileGrouping = SourceFileGroupingBuilder.Build(this.fileResolver, currentModuleDispatcher, this.workspace, inputUri, configuration);
            var originalModulesToRestore = sourceFileGrouping.ModulesToRestore;

            // RestoreModules() does a distinct but we'll do it also to prevent duplicates in processing and logging
            var modulesToRestoreReferences = currentModuleDispatcher.GetValidModuleReferences(sourceFileGrouping.ModulesToRestore, configuration)
                .Distinct()
                .OrderBy(key => key.FullyQualifiedReference);

            // restore is supposed to only restore the module references that are syntactically valid
            await currentModuleDispatcher.RestoreModules(configuration, modulesToRestoreReferences);

            ...


```

Another idea would be to add a new ModuleRestoreStatus for the "cache deprecation" cleanup phase. Then we could properly handle the transition from one and another instead of hijacking the restore status ? It does implies a more work though.

Reply:
> My other worry with doing the calculations like this is that the logic could get out of sync with the logic called by `Build()`.
> 
> Could we just pass a flag into `Build()` so it ignores restore status and always acts like the module is not restored?

@majastrz I gave it a second look and though, thinking it would go faster like you mentioned since we were almost there.

If we do this, we really start to put logic in `SourceFileGroupingBuilder` that take decisions away from `ModuleDispatcher`.

1. First it would override the `restoreStatus` 
2. It needs also to override `restoreErrorBuilder` which would have returned an `ErrorDiagnostic` via an `out` `ModuleRequiresRestore`.

The code is written on my machine and it works, it really depends how we want the concerns to be separated.

![image](https://user-images.githubusercontent.com/1054412/164461275-cd777da9-4b08-429e-99ee-4be01c0b0c0e.png)

I think we can deliver the force restore that way and create a new issue to rethink a bit the interaction and extensibility for components like `ModuleDispatcher` & `SourceFileGroupingBuilder` to prevent situations like this in the future.

If you give me a green light on this, I will commit the new code I have on my machine that support you suggestion and screenshot above. It removes and revert the ForceRestoreModuleDispatcher concept introduced previously.

IMO we should have been able to add this kind of functionality easily through aggregation/composition without touching much of the existing code, just extending/replacing behaviors.